vmm/src/vm.rs

```diff
@@ -295,6 +298,156 @@ impl VmState {
+    fn log_debug_ioport(&self, code: u8) {
+        let elapsed = self.timestamp.elapsed();
+
+        debug!(
+            "[{} code 0x{:x}] {}.{:>06} seconds",
+            DebugIoPortRange::from_u8(code),
+            code,
+            elapsed.as_secs(),
+            elapsed.as_micros()
+        );
+    }
+}
+
+impl VmmOps for VmOps {
+    fn guest_mem_write(&self, buf: &[u8], gpa: u64) -> hypervisor::vm::Result<usize> {
+        let guest_memory = self.memory_manager.lock().unwrap().guest_memory().memory();
+        guest_memory
+            .write(buf, GuestAddress(gpa))
+            .map_err(|e| HypervisorVmError::GuestMemWrite(e.into()))
+    }
+
+    fn guest_mem_read(&self, buf: &mut [u8], gpa: u64) -> hypervisor::vm::Result<usize> {
+        let guest_memory = self.memory_manager.lock().unwrap().guest_memory().memory();
+        guest_memory
+            .read(buf, GuestAddress(gpa))
+            .map_err(|e| HypervisorVmError::GuestMemRead(e.into()))
+    }
+
+    fn mmio_read(&self, addr: u64, data: &mut [u8]) -> hypervisor::vm::Result<()> {
+        if let Err(e) = self
+            .device_manager
+            .lock()
+            .unwrap()
+            .mmio_bus()
+            .read(addr, data)
+        {
+            if let vm_device::BusError::MissingAddressRange = e {
+                warn!("Guest MMIO read to unregistered address 0x{:x}", addr);
+            }
+        }
+        Ok(())
+    }
+
+    fn mmio_write(&self, addr: u64, data: &[u8]) -> hypervisor::vm::Result<()> {
+        if let Err(e) = self
+            .device_manager
+            .lock()
+            .unwrap()
+            .mmio_bus()
+            .write(addr, data)
+        {
+            if let vm_device::BusError::MissingAddressRange = e {
+                warn!("Guest MMIO write to unregistered address 0x{:x}", addr);
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn pio_read(&self, addr: u64, data: &mut [u8]) -> hypervisor::vm::Result<()> {
+        if let Err(e) = self
+            .device_manager
+            .lock()
+            .unwrap()
+            .io_bus()
+            .read(addr, data)
+        {
+            if let vm_device::BusError::MissingAddressRange = e {
+                warn!("Guest PIO read to unregistered address 0x{:x}", addr);
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn pio_write(&self, addr: u64, data: &[u8]) -> hypervisor::vm::Result<()> {
+        if addr == DEBUG_IOPORT as u64 && data.len() == 1 {
+            self.log_debug_ioport(data[0]);
+        }
+
+        if let Err(e) = self
+            .device_manager
+            .lock()
+            .unwrap()
+            .io_bus()
+            .write(addr, data)
+        {
+            if let vm_device::BusError::MissingAddressRange = e {
+                warn!("Guest PIO write to unregistered address 0x{:x}", addr);
+            }
+        }
+        Ok(())
+    }
+}
```

Comment:
This has been written in order to keep the previous behavior, instead of propagating the error, which would have been a change.